[ iOS ] css3/filters/backdrop-filter-page-scale.html is a flaky image failure.

mattwoodrow opened the pull request:
#### b7e4c4a702a5b14d31790daf0a89850516b5ddf2
<pre>
[ iOS ] css3/filters/backdrop-filter-page-scale.html is a flaky image failure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=280260">https://bugs.webkit.org/show_bug.cgi?id=280260</a>
&lt;<a href="https://rdar.apple.com/134269886">rdar://134269886</a>&gt;

Reviewed by Wenson Hsieh.

The main change here is extending the fix in 258521@main to apply to the X axis
as well.

This test applies a page scale (with no offset) in the WebProcess (via a
WebPageProxy).  When this reaches WkWebViewIOS in the following
RemoteLayerTreeTransaction, it applies the same zoomScale, and zooms towards the
middle in the X axis. This leaves the left 200px offscreen (which includes all
the test content).  updateVisibleContentRects is then sent back to the
WebProcess with the new scroll offset, which calls setPageScaleFactor a second
time, with the scroll offset.

This test regressed by making it more async, which gives time for this multiple
transaction scrolling to take place, and then the test content is sometimes
offscreen.

This also fixes the TestRunner setPageScaleFactor setter to wait until a
rendering update has been done with the new page scale, before allowing tests to
complete.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Source/WebKit/UIProcess/WebPageProxyTesting.cpp:
(WebKit::WebPageProxyTesting::setPageScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
(WebKit::WebPageTesting::setPageScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPageTesting.h:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.messages.in:

Canonical link: <a href="https://commits.webkit.org/284197@main">https://commits.webkit.org/284197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1003b5305a69075e601b35b05ee066f64dc6157

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68705 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48097 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21364 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72775 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19850 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55893 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19666 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54762 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13193 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59318 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35226 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40609 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16743 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18208 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62558 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17091 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74468 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12676 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16338 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62243 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-translate.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12716 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59399 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62271 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15253 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10235 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3848 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43898 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44972 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46166 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44714 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1006,13 +1006,14 @@ - (void)_updateScrollViewForTransaction:(const WebKit::RemoteLayerTreeTransactio
+        // scroll position to the top edge of the content,
         // instead of the center (which UIScrollView does by default).
         CGFloat contentOffsetY = [_scrollView contentOffset].y * layerTreeTransaction.pageScaleFactor() / [_scrollView zoomScale];
+        CGFloat contentOffsetX = [_scrollView contentOffset].x * layerTreeTransaction.pageScaleFactor() / [_scrollView zoomScale];
```
Nit - maybe we can pull `layerTreeTransaction.pageScaleFactor() / [_scrollView zoomScale]` out into a separate local variable?

mattwoodrow replied:
Good call, done!

mattwoodrow committed:
[ iOS ] css3/filters/backdrop-filter-page-scale.html is a flaky image failure.
https://bugs.webkit.org/show_bug.cgi?id=280260
<rdar://134269886>

Reviewed by Wenson Hsieh.

The main change here is extending the fix in 258521@main to apply to the X axis
as well.

This test applies a page scale (with no offset) in the WebProcess (via a
WebPageProxy).  When this reaches WkWebViewIOS in the following
RemoteLayerTreeTransaction, it applies the same zoomScale, and zooms towards the
middle in the X axis. This leaves the left 200px offscreen (which includes all
the test content).  updateVisibleContentRects is then sent back to the
WebProcess with the new scroll offset, which calls setPageScaleFactor a second
time, with the scroll offset.

This test regressed by making it more async, which gives time for this multiple
transaction scrolling to take place, and then the test content is sometimes
offscreen.

This also fixes the TestRunner setPageScaleFactor setter to wait until a
rendering update has been done with the new page scale, before allowing tests to
complete.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Source/WebKit/UIProcess/WebPageProxyTesting.cpp:
(WebKit::WebPageProxyTesting::setPageScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
(WebKit::WebPageTesting::setPageScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPageTesting.h:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.messages.in:

Canonical link: https://commits.webkit.org/284197@main
